Core: Add Yarn 2 compatibility

gaetanmaisse opened the pull request:
Issue: 
Part of #9527

🏗 Some changes will also be needed on [CRA preset](https://github.com/storybookjs/presets/blob/65c2968fe5389be3677643278a0d5fa4537fac40/packages/preset-create-react-app/src/helpers/getReactScriptsPath.ts) to make it compatible.

## What I did

 - Move Webpack default `outputDir` outside `@storybook/core` package
 - Improve management of cache directory
 - Update Webpack configuration to handle PnP correctly
 - Add missing or fix implicit transitive peerDependencies 

## How to test

 - Checkout the branch locally, compile `@storybook/core`
 - Follow https://github.com/crubier/yarnv2-storybook 
 - In the `package.json` file add your local version of `@storybook/core`:
```json
  "resolutions": {
    "@storybook/core": "portal:/PATH_TO_SB/lib/core"
  }
```
 - `yarn install` 
 - `yarn storybook` and 🎉🎉 

--
Does this need a new example in the kitchen sink apps?
--> I think so, but for now, I didn't find a way to use Yarn v2 on some project of our Yarn v1 workspace.

gaetanmaisse committed:
feat(core): create a function to resolve path of a file/dir in storybook cache directory

`find-cache-dir` is not used anymore because it returns `undefined` when there is no `node_modules` folder at project root, as in projects managed with Yarn 2.

gaetanmaisse committed:
refactor(core): avoid duplication of code used to get path of files/dirs in cache

gaetanmaisse committed:
feat(core): move webpack default output from `storybook/core` own package folder to SB cache folder

Using Yarn 2 packages are read only, and more globally it's better to not alter the package with its own execution, so the output is moved outside the packages.

gaetanmaisse committed:
feat(core): update webpack config to handle Yarn PnP mode out of the box

Add PnpWebpackPlugin in manager and preview webpack configurations

shilman commented:
Love this @gaetanmaisse !!!

gaetanmaisse commented:
@shilman maybe `cra-kitchen-sink` or `html-kitchen-sink` because they both have a simple configuration with no presets. What's your point of view?  

I still need to do a few more tests before setting this PR as ready but I think there will not be many more code changes on `@storybook/core` side 😉. Also, I started to look at the `presets` side (as CRA TS is not working for now): 

- `@storybook/preset-typescript` should work without any change. I will test it. 
- `@storybook/preset-create-react-app` will need some updates. It's currently resolving `react-script` in `node_modules/.bin/` directory (see [here](https://github.com/storybookjs/presets/blob/65c2968fe5389be3677643278a0d5fa4537fac40/packages/preset-create-react-app/src/helpers/getReactScriptsPath.ts)), so I need to add one more check for SB PnP user and access webpack config file directly in the .zip file with [PnP API](https://yarnpkg.com/advanced/pnpapi#accessing-the-files).

shilman commented:
Yeah I’d say do both! CRA because it’s kind of “special” and one of the others to have a “normal” example. Would be nice to have one that tests typescript support too. If this is the future of yarn, doing a few should be fine. As long as we also have a few old-style examples around too

gaetanmaisse committed:
chore: add missing or fix implicit transitive peerDependencies

Yarn 2 introduce an error for this during `yarn install`: YN0002 - MISSING_PEER_DEPENDENCY

For details see https://dev.to/arcanis/implicit-transitive-peer-dependencies-ed0

TL;DR: If you write a package that depends on Foo, and if Foo has a peer dependency, then you must provide it in either of the dependencies or peerDependencies fields. You won't "implicitly inherit" the peer dependencies declared in Foo.

gaetanmaisse commented:
I open this PR because with these commits everything should be OK for Yarn 2 & Storybook compatibility on simple projects. ⚠️  For now `@storybook/preset-create-react-app` is not working and I haven't tried all addons.

@shilman I didn't find a way to use Yarn 2 on only some projects of our Yarn v1 workspace. As soon as this PR will be merged and released I will add an e2e test on [Yarn 2 repo](https://github.com/yarnpkg/berry#current-status) matching the scenario given in the issue.


Then I will work on: 
- Updating `@storybook/preset-create-react-app`. it's currently resolving `react-script` in `node_modules/.bin/` directory (see [here](https://github.com/storybookjs/presets/blob/65c2968fe5389be3677643278a0d5fa4537fac40/packages/preset-create-react-app/src/helpers/getReactScriptsPath.ts)), so I need to add one more check for SB PnP user and access webpack config file directly in the .zip file with [PnP API](https://yarnpkg.com/advanced/pnpapi#accessing-the-files).
 - Checking that all addons are working with the alpha of SB 6.0 that will release this PR.
 - Finding a way of adding examples with Yarn 2 inside SB monorepo. Maybe creating some GitHub actions as it is done on Yarn 2 repo can be a good idea because it really mimics the users' behaviors: https://github.com/yarnpkg/berry/blob/master/.github/workflows/e2e-jest-workflow.yml

shilman commented:
@gaetanmaisse This is really awesome work -- will be so great to announce this as part of the 6.0 release. 🤩🤩🤩

I'm not tracking the yarn 2 situation closely, but based on my limited understanding, we will need to support both yarn 1 and 2 for the foreseeable future. Therefore here's my strawman proposal, assuming that yarn 2 will be the **primary** use case moving forward, and yarn 1 will be the **secondary**:

1) Let's convert the monorepo to yarn 2.
2) Let's use the `storybookjs/deprecated-addons` repo as a yarn 1 test case.

What do you think? Alternatively, if I got the primary/secondary cases wrong, we can flip the proposal.

shilman commented on addons/contexts/package.json:
```diff
@@ -41,7 +41,20 @@
     "rax": "*",
     "react": "*",
-    "vue": "*"
+    "vue": "*",
```
@gaetanmaisse this PR somehow got removed: https://github.com/storybookjs/storybook/pull/7675

Do you think I should reapply that first before this PR? Would it change this part of this PR?

gaetanmaisse replied:
I think we can keep it as it is and as it will be with this PR, from my point of view it's more semantically correct to have `preact`, `rax` and `vue` listed as peerDep as we are importing them in the source code of `addon-context` (in `addons/contexts/preview/framework`). Adding the `peerDependenciesMeta` info will prevent any npm/yarn warning about missing dep which was the main point of https://github.com/storybookjs/storybook/issues/7625 and https://github.com/storybookjs/storybook/issues/7625#issuecomment-517627982.

shilman replied:
OK cool

gaetanmaisse commented:
@shilman I like your proposal! I will add to my todo list to investigate it 🕵 . 

It is maybe a bit too ambitious to migrate our workspace to Yarn 2 right now because: 
 - we don't have any feedback of big projects using it
 - I'm only playing with it on small personal projects for a few days

Anyway, I will take a look at what is possible and I will share my feedback and thoughts 😉

shilman reviewed (approved):
LGTM. Could use confirmation from @ndelangen or somebody else who's more familiar with the situation.
